File: 2022.01.30/javascript/1654.js
const input = require('fs').readFileSync('/dev/stdin').toString().split('\n');
const [k,n] = input[0].split(' ');
let arr = [];
for (let i = 1 ; i<input.length-1 ; i++){
  arr.push(Number(input[i]));
}
let result = 0;
const search = (minNum,maxNum) => {
  if (minNum > maxNum){
    return;
  }
  let halfNum = minNum + Math.floor((maxNum-minNum)/2);
  const cnt = arr.reduce((cnt,cur) => {return cnt+Math.floor(cur/halfNum)},0)
  if (cnt>=n){
    result = halfNum;
    search(halfNum+1,maxNum)
  }
  if (cnt<n){
    search(minNum,halfNum-1);
  }
}
search(0,arr.reduce((prev,cur)=>Math.max(prev,cur)));
console.log(result);

File: 2022.01.30/python/1654.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
k,n = map(int,input().split())
arr = [int(input()) for _ in range(k)]
result = 0
def search(min_num,max_num):
    global result
    if min_num>max_num : return
    half_num = min_num + (max_num-min_num)//2
    cnt = 0
    for num in arr:
        cnt += num//half_num
    if cnt>=n:
        result = half_num
        search(half_num+1,max_num)
    else:
        search(min_num,half_num-1)
search(1,max(arr))
print(result)